File: java/client/src/main/java/glide/api/models/commands/scan/ZScanOptions.java
/** Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0 */
package glide.api.models.commands.scan;

import glide.api.commands.SortedSetBaseCommands;
import glide.api.models.GlideString;
import glide.utils.ArgsBuilder;
import java.util.Arrays;
import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.experimental.SuperBuilder;

/**
 * Optional arguments for {@link SortedSetBaseCommands#zscan(String, String, ZScanOptions)}.
 *
 * @see <a href="https://valkey.io/commands/zscan/">valkey.io</a>
 */
@SuperBuilder
@EqualsAndHashCode(callSuper = false)
public class ZScanOptions extends BaseScanOptions {

    /** Option string to include in the ZSCAN command when scores are not included. */
    public static final String NO_SCORES_API = "NOSCORES";

    /**
     * When set to true, the command will not include scores in the results. This option is available
     * from Valkey version 8.0.0 and above.
     */
    @Builder.Default protected boolean noScores = false;

    @Override
    public String[] toArgs() {
        return Arrays.stream(toGlideStringArgs()).map(GlideString::getString).toArray(String[]::new);
    }

    /**
     * Creates the arguments to be used in <code>ZSCAN</code> commands.
     *
     * @return a GlideString array that holds the options and their arguments.
     */
    @Override
    public GlideString[] toGlideStringArgs() {
        ArgsBuilder builder = new ArgsBuilder();

        // Add options from the superclass
        GlideString[] superArgs = super.toGlideStringArgs();
        for (GlideString arg : superArgs) {
            builder.add(arg);
        }

        // Add the noScores option if applicable
        if (noScores) {
            builder.add(NO_SCORES_API);
        }

        return builder.toArray();
    }
}

File: go/api/options/zscan_options.go
// Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

package options

// This struct represents the optional arguments for the ZSCAN command.
type ZScanOptions struct {
	BaseScanOptions
	noScores bool
}

func NewZScanOptionsBuilder() *ZScanOptions {
	return &ZScanOptions{}
}

// SetNoScores sets the noScores flag for the ZSCAN command.
// If this value is set to true, the ZSCAN command will be called with NOSCORES option.
// In the NOSCORES option, scores are not included in the response.
func (zScanOptions *ZScanOptions) SetNoScores(noScores bool) *ZScanOptions {
	zScanOptions.noScores = noScores
	return zScanOptions
}

// SetMatch sets the match pattern for the ZSCAN command.
func (zScanOptions *ZScanOptions) SetMatch(match string) *ZScanOptions {
	zScanOptions.BaseScanOptions.SetMatch(match)
	return zScanOptions
}

// SetCount sets the count of the ZSCAN command.
func (zScanOptions *ZScanOptions) SetCount(count int64) *ZScanOptions {
	zScanOptions.BaseScanOptions.SetCount(count)
	return zScanOptions
}

func (options *ZScanOptions) ToArgs() ([]string, error) {
	args := []string{}
	baseArgs, err := options.BaseScanOptions.ToArgs()
	args = append(args, baseArgs...)

	if options.noScores {
		args = append(args, NoScores)
	}
	return args, err
}
